### PR TITLE
condition: add new ConditionKernelModuleLoaded=

### PR DESCRIPTION
**Pull request by poettering**
This introduces a new unit condition check: that matches if a specific kmod module is loaded. This should be generally useful, but there's one usecase in particular: we can optimize modprobe@.service with this and avoid forking out a bunch of modprobe requests during boot for the same kmods.

Checking if a kernel module is loaded is more complicated than just checking if /sys/module/$MODULE/ exists, since kernel modules typically take a while to initialize and we must check that this is complete (by checking if the sysfs attr "initstate" is "live").

### CODE REVIEW

File: src/shared/condition.c [hunk 1]
```diff
@@ -1155,6 +1155,59 @@ static int condition_test_psi(Condition *c, char **env) {
...
+        }
+
+        _cleanup_free_ char *p = path_join("/sys/module/", normalized);
+        if (!p)
+                return log_oom_debug();
+
+        _cleanup_close_ int dir_fd = open(p, O_DIRECTORY|O_CLOEXEC);
+        if (dir_fd < 0) {
+                if (errno == ENOENT) {
+                        log_debug("%s does not exist, kernel module '%s' not loaded.", p, normalized);
```
Comment (YHNdnzj) [inline src/shared/condition.c]: _errno to submit errno to journal?

File: src/shared/condition.c [hunk 2]
```diff
@@ -1155,6 +1155,59 @@ static int condition_test_psi(Condition *c, char **env) {
...
+                        return false;
+                }
+
+                return log_debug_errno(errno, "Failed to open directory '%s/': %m", p);
+        }
+
+        _cleanup_free_ char *initstate = NULL;
+        r = read_virtual_file_at(dir_fd, "initstate", SIZE_MAX, &initstate, NULL);
+        if (r == -ENOENT) {
+                log_debug("%s exists but %s/initstate does not, kernel modules is built-in, hence loaded.", p, p);
```
Comment (YHNdnzj) [inline src/shared/condition.c]: ditto

File: src/shared/condition.c [hunk 3]
```diff
@@ -1155,6 +1155,59 @@ static int condition_test_psi(Condition *c, char **env) {
...
+        if (!filename_is_valid(normalized)) {
+                log_debug("Kernel module name '%s' is not valid, hence reporting it to not be loaded.", normalized);
+                return false;
+        }
+
+        _cleanup_free_ char *p = path_join("/sys/module/", normalized);
+        if (!p)
+                return log_oom_debug();
+
+        _cleanup_close_ int dir_fd = open(p, O_DIRECTORY|O_CLOEXEC);
```
Comment (YHNdnzj) [inline src/shared/condition.c]: This can be O_PATH?

File: src/shared/condition.c (hunk 2 above)
Comment (YHNdnzj) [inline src/shared/condition.c]:
```suggestion
                log_debug("%s exists but %s/initstate does not, kernel module '%s' is built-in.", p, p, normalized);
```

File: src/shared/condition.c [hunk 4]
```diff
@@ -1155,6 +1155,59 @@ static int condition_test_psi(Condition *c, char **env) {
...
+        }
+
+        _cleanup_free_ char *initstate = NULL;
+        r = read_virtual_file_at(dir_fd, "initstate", SIZE_MAX, &initstate, NULL);
+        if (r == -ENOENT) {
+                log_debug("%s exists but %s/initstate does not, kernel modules is built-in, hence loaded.", p, p);
+                return true;
+        }
+        if (r < 0)
+                return log_debug_errno(r, "Failed to open %s/initstate: %m", p);
```
Comment (YHNdnzj) [inline src/shared/condition.c]:
```suggestion
                return log_debug_errno(r, "Failed to open '%s/initstate': %m", p);
```

File: src/shared/condition.c [hunk 5]
```diff
@@ -1155,6 +1155,59 @@ static int condition_test_psi(Condition *c, char **env) {
...
+
+        delete_trailing_chars(initstate, WHITESPACE);
+
+        if (!streq(initstate, "live")) {
```
Comment (YHNdnzj) [inline src/shared/condition.c:1202]: streq_skip_trailing_chars

**Review by YHNdnzj — APPROVED** [review]

File: src/shared/condition.c (hunk 5 above)
Comment (yuwata) [reply src/shared/condition.c:1202]: `streq()` should be OK, otherwise the log message below may contain trailing white spaces.

**Review by yuwata — APPROVED** [review]
LGTM. Nice.

File: src/shared/condition.c (hunk 5 above)
Comment (YHNdnzj) [reply src/shared/condition.c:1202]: Ah, indeed. Please ignore then.

**Pull request merged**